From 904a65fc0bcdab84c09facc9e912d746597afa3e Mon Sep 17 00:00:00 2001
From: Wujian Sun <wujian.sun_1@nxp.com>
Date: Mon, 15 Dec 2025 10:01:04 +0800
Subject: [PATCH] Fix webgl regression after angle upgrade

conformance/ogles/GL/functions/functions_057_to_064.html: 16 tests failed
conformance/ogles/GL/functions/functions_065_to_072.html: 4 tests failed

Revert "Translator: Fix output-variable-init vs after-main declarations"

This reverts commit dba77589913e6880ac5dad665f1c2930692a8867.

Upstream-Status: Inappropriate [i.MX specific]

Signed-off-by: Wujian Sun <wujian.sun_1@nxp.com>
---
 src/compiler/translator/Compiler.cpp          | 26 -------------------
 src/compiler/translator/IntermNode.cpp        |  5 ++--
 src/compiler/translator/IntermNode.h          |  2 +-
 .../MonomorphizeUnsupportedFunctions.cpp      |  2 +-
 src/tests/gl_tests/GLSLTest.cpp               | 11 --------
 5 files changed, 5 insertions(+), 41 deletions(-)

diff --git a/third_party/angle/src/compiler/translator/Compiler.cpp b/third_party/angle/src/compiler/translator/Compiler.cpp
index aa3e57f7d1..e54f44ebb8 100644
--- a/third_party/angle/src/compiler/translator/Compiler.cpp
+++ b/third_party/angle/src/compiler/translator/Compiler.cpp
@@ -1855,32 +1855,6 @@ bool TCompiler::useAllMembersInUnusedStandardAndSharedBlocks(TIntermBlock *root)
 
 bool TCompiler::initializeOutputVariables(TIntermBlock *root)
 {
-    // Place `main` at the end of the shader if not already.  If a variable is declared after main,
-    // main cannot reference it.
-    {
-        const TIntermSequence *original = root->getSequence();
-        TIntermSequence reordered;
-        TIntermNode *main = nullptr;
-
-        for (TIntermNode *node : *original)
-        {
-            TIntermFunctionDefinition *function = node->getAsFunctionDefinition();
-            if (function != nullptr && function->getFunction()->isMain())
-            {
-                ASSERT(main == nullptr);
-                main = node;
-            }
-            else
-            {
-                reordered.push_back(node);
-            }
-        }
-        ASSERT(main != nullptr);
-        reordered.push_back(main);
-
-        root->replaceAllChildren(std::move(reordered));
-    }
-
     InitVariableList list;
 
     for (TIntermNode *node : *root->getSequence())
diff --git a/third_party/angle/src/compiler/translator/IntermNode.cpp b/third_party/angle/src/compiler/translator/IntermNode.cpp
index e8faeb41e6..5d39a4a7e6 100644
--- a/third_party/angle/src/compiler/translator/IntermNode.cpp
+++ b/third_party/angle/src/compiler/translator/IntermNode.cpp
@@ -467,9 +467,10 @@ bool TIntermBlock::replaceChildNode(TIntermNode *original, TIntermNode *replacem
     return replaceChildNodeInternal(original, replacement);
 }
 
-void TIntermBlock::replaceAllChildren(TIntermSequence &&newStatements)
+void TIntermBlock::replaceAllChildren(const TIntermSequence &newStatements)
 {
-    mStatements = std::move(newStatements);
+    mStatements.clear();
+    mStatements.insert(mStatements.begin(), newStatements.begin(), newStatements.end());
 }
 
 size_t TIntermFunctionPrototype::getChildCount() const
diff --git a/third_party/angle/src/compiler/translator/IntermNode.h b/third_party/angle/src/compiler/translator/IntermNode.h
index 358c5bf07e..283401fd33 100644
--- a/third_party/angle/src/compiler/translator/IntermNode.h
+++ b/third_party/angle/src/compiler/translator/IntermNode.h
@@ -700,7 +700,7 @@ class TIntermBlock : public TIntermNode, public TIntermAggregateBase
     size_t getChildCount() const final;
     TIntermNode *getChildNode(size_t index) const final;
     bool replaceChildNode(TIntermNode *original, TIntermNode *replacement) override;
-    void replaceAllChildren(TIntermSequence &&newStatements);
+    void replaceAllChildren(const TIntermSequence &newStatements);
 
     // Only intended for initially building the block.
     void appendStatement(TIntermNode *statement);
diff --git a/third_party/angle/src/compiler/translator/tree_ops/MonomorphizeUnsupportedFunctions.cpp b/third_party/angle/src/compiler/translator/tree_ops/MonomorphizeUnsupportedFunctions.cpp
index 10ad06b7e6..d65263ff83 100644
--- a/third_party/angle/src/compiler/translator/tree_ops/MonomorphizeUnsupportedFunctions.cpp
+++ b/third_party/angle/src/compiler/translator/tree_ops/MonomorphizeUnsupportedFunctions.cpp
@@ -535,7 +535,7 @@ void SortDeclarations(TIntermBlock *root)
     replacement.insert(replacement.end(), functionDefs.begin(), functionDefs.end());
 
     // Replace root's sequence with |replacement|.
-    root->replaceAllChildren(std::move(replacement));
+    root->replaceAllChildren(replacement);
 }
 
 bool MonomorphizeUnsupportedFunctionsImpl(TCompiler *compiler,
diff --git a/third_party/angle/src/tests/gl_tests/GLSLTest.cpp b/third_party/angle/src/tests/gl_tests/GLSLTest.cpp
index c7da2e8f3a..e9942e5a65 100644
--- a/third_party/angle/src/tests/gl_tests/GLSLTest.cpp
+++ b/third_party/angle/src/tests/gl_tests/GLSLTest.cpp
@@ -11139,17 +11139,6 @@ void main()
     EXPECT_NE(shader, 0u);
 }
 
-// Test that output variables declared after main work in combination with initOutputVariables
-// (which is enabled on WebGL).
-TEST_P(WebGLGLSLTest, OutputAfterMain)
-{
-    constexpr char kVS[] = R"(void main(){}
-varying float r;)";
-
-    GLuint shader = CompileShader(GL_VERTEX_SHADER, kVS);
-    EXPECT_NE(shader, 0u);
-}
-
 // Test that clamp applied on non-literal indices is correct on es 100 shaders.
 TEST_P(GLSLTest, ValidIndexClampES100)
 {
-- 
2.34.1

